policy/src/main/java/org/apache/brooklyn/policy/ha/ElectPrimaryEffector.java

```diff
@@ -0,0 +1,440 @@
+                }
+                
+                if (newPrimary.equals(currentActive)) {
+                    // If there is a primary and it is unchanged, the effector will end.
+                    return MutableMap.of("code", ResultCode.PRIMARY_UNCHANGED, "message", "No change required", "primary", newPrimary);
+                }
+                
+                log.info("Detected new primary "+newPrimary+" at "+entity()+" (previously had "+currentActive+")");
+//                If a new primary is detected, the effector will:
+//                    * set the local entity to the STARTING state
+//                    * clear any "primary-election" problem
+//                    * publish the new primary in a sensor called `primary` (or the sensor set in `primary.sensor.name`)
+//                    * cancel any other ongoing promote calls, and if there is an ongoing demote call on the entity being promoted, cancel that also
+//                    * in parallel
+//                        * invoke `promote` (or the effector called `primary.promote.effector.name`) on the local entity or the entity being promoted
+//                        * invoke `demote` (or the effector called `primary.promote.effector.name`) on the local entity or the entity being demoted, if an entity is being demoted
+//                    * set service up true
+//                    * set the local entity to the RUNNING state
+                
+                boolean wasRunning = entity().sensors().get(Attributes.SERVICE_STATE_ACTUAL) == Lifecycle.RUNNING;
+                if (wasRunning) {
+                    log.debug("Transititioning "+entity()+" to starting while promoting/demoting");
+                    ServiceStateLogic.setExpectedState(entity(), Lifecycle.STARTING);
+                }
+                ServiceProblemsLogic.clearProblemsIndicator(entity(), "primary");
+                entity().sensors().set(Sensors.newSensor(Entity.class, params.get(PRIMARY_SENSOR_NAME)), newPrimary);
+                try {
+                    // TODO cancel other promote/demote calls
+        
+                    promoteAndDemote(params, currentActive, newPrimary);
+                    
+                    log.debug("Promoted/demoted primary for "+entity()+", now setting service up "+(wasRunning ? "and running" : "(but not setting as 'running' because it wasn't 'running' before)"));
+                    entity().sensors().set(Attributes.SERVICE_UP, true);
+                    if (wasRunning) ServiceStateLogic.setExpectedState(entity(), Lifecycle.RUNNING);
+                    
+                } catch (Exception e) {
+                    Exceptions.propagateIfFatal(e);
+                    log.debug("Error promoting/demoting primary for "+entity()+" (rethrowing): "+e);
+                    ServiceProblemsLogic.updateProblemsIndicator(entity(), "primary", e);
```

Comment:
I'd set the problems indicator as `e.toString()`. Otherwise we end up serializing/persisting the exception instance.

Reply:
+1 - using `Exceptions.collapseText(e)`